tools/lib/dependencies.js

```diff
@@ -0,0 +1,84 @@
+const DependencyResolver = require('dependency-resolver');
+
+/**
+ * Reorders languages, moving dependencies above the languages
+ * that depend on them.
+ *
+ * @param {array<Language>} languages list of languages to be reordered
+ * @returns {array<Language>} ordered list of languages
+*/
+
+const reorderDependencies = (languages) => {
+  let resolver = new DependencyResolver();
+  for (let lang of languages) {
+    resolver.add(lang.name);
+    for (let required of lang.requires) {
+      resolver.setDependency(lang.name, required);
+    }
+  }
+  return resolver.sort().map((name) =>
+    languages.find((l) => l.name == name)
+  );
+}
+
+/**
+ * Filters languages by group (common, etc)
+ *
+ * Groups are currently defined inside each language file with `Category`
+ *
+ * @param {array<Language>} languages full list of languages
+ * @returns {array<Language>} filtered list
+*/
+
+const languagesByGroup = (languages, groupName) => {
+  return languages.filter((el) => el.categories.includes(groupName))
+}
+const isGroupName = (name) => name[0] == ":"
+
+
+/**
+ * Filters languages by a list of languages or categories
+ *
+ * This also resolves any requires and dependency ordering issues.
+ * The returned list can be registered sequentially and should "just work".
+ *
+ * @param {array<Lanuage>} allLanguages - full list of languages
+ * @param {array<name|group_name>} includes - which languages or groups to include
+ *   example: ":common elixir ruby"
+ * @returns {array<Language>} filtered list if languages
+*/
+const filter = (allLanguages, includes) => {
+  if (includes==undefined || includes.length==0)
+    return reorderDependencies(allLanguages);
+
+  let languages = [];
+  for (let lang of includes) {
+    if (isGroupName(lang)) {
+      let group = lang.substr(1)
+      languages = languages.concat(languagesByGroup(allLanguages,group))
+    } else {
+      let found = allLanguages.find((el) => el.name == lang )
+      if (found)
+        languages.push(found)
+      else {
+        console.error(`[ERROR] Language '${lang}' could not be found.`)
+        process.exit(1)
+      }
+    }
+  }
+
+  // resolve requires
+  for (let lang of languages) {
+    lang.requires.forEach(needed => {
+      if (!languages.find((el) => el.name == needed)) {
+        console.info(`[INFO] Adding ${needed}... ${lang.name} requires ${needed}.`)
+        languages.push(allLanguages.find((el) => el.name == needed))
```

Comment:
No, it's simply adding requires.  Duplicates would be resolved by the dependency resolver.

Reply:
Then why do you need this check in line 75?
```js
 if (!languages.find((el) => el.name == needed)) {
```

Reply:
There is no need to add a requirement twice... one task shouldn't NOT do something entirely obvious just because another later MIGHT take care of it... **when adding dependencies there is no need to add a dependent that already exists.** Even if it might get cleaned up later.

I see no issue here.

Reply:
Perhaps if it was all a single function, but it's not... and also the logging would not be nearly so helpful if it just added things over and over even if they already existed.